File: android/src/main/kotlin/com/emarsys/emarsys_sdk/command/mobileengage/inbox/FetchMessagesCommand.kt
package com.emarsys.emarsys_sdk.command.mobileengage.inbox

import com.emarsys.Emarsys
import com.emarsys.emarsys_sdk.command.EmarsysCommand
import com.emarsys.emarsys_sdk.command.ResultCallback
import com.emarsys.emarsys_sdk.mapper.InboxResultMapper
import com.emarsys.mobileengage.api.inbox.InboxResult

class FetchMessagesCommand(private val inboxResultMapper: InboxResultMapper) : EmarsysCommand {
    override fun execute(parameters: Map<String, Any?>?, resultCallback: ResultCallback) {

        Emarsys.messageInbox.fetchMessages { result ->
            if (result.errorCause != null) {
                resultCallback.invoke(null, result.errorCause)
                return@fetchMessages
            }
            resultCallback.invoke((inboxResultMapper.map(result.result as InboxResult)), null)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        return true
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }

}

File: ios/Classes/Wrapper/Commands/Inbox/FetchMessagesCommand.swift
//
//  Created by Emarsys on 2021. 08. 18..
//

import EmarsysSDK

class FetchMessagesCommand: EmarsysCommandProtocol {
    
    var inboxMapper: InboxMapper
    
    init(inboxMapper: InboxMapper) {
        self.inboxMapper = inboxMapper
    }
    
    func execute(arguments: [String : Any]?, resultCallback: @escaping ResultCallback) {
        Emarsys.messageInbox.fetchMessages(resultBlock: { result, error in
            if let res = result {
                let messages: [EMSMessage] = res.messages
                if let messageDicts = self.inboxMapper.map(messages) {
                    resultCallback(["success": messageDicts])
                } else {
                    resultCallback(["success": []])
                }
            } else if let e = error {
                resultCallback(["error": e])
            }
        })
    }
}
